app/models/user.rb

```diff
@@ -163,31 +163,43 @@ def course_student?
   end
 
   def nonvisitor?(course)
-    role(course) != CoursesUsers::Roles::VISITOR_ROLE
+    return true if admin?
+    roles = course_roles(course)
+    roles.size > 1 || (roles.size == 1 && roles.first != CoursesUsers::Roles::VISITOR_ROLE)
+  end
+
+  # returns an array of roles a user has in a given course
+  def course_roles(course)
+    user_course_roles = course.courses_users.where(user_id: id).order('role DESC').pluck(:role)
+    return user_course_roles
   end
 
-  def role(course)
-    # If this is a new course, grant permissions.
-    return CoursesUsers::Roles::INSTRUCTOR_ROLE if course.nil?
-    # Give admins the instructor permissions.
-    return CoursesUsers::Roles::INSTRUCTOR_ROLE if admin?
+  def highest_role(course)
+    roles = course_roles(course)
 
-    course_user = course.courses_users.where(user_id: id).order('role DESC').first
-    return course_user.role unless course_user.nil?
+    return CoursesUsers::Roles::VISITOR_ROLE if roles.empty?
 
-    # User is in visitor role, if no other role found.
-    CoursesUsers::Roles::VISITOR_ROLE
+    roles.first
+  end
+
+  def editing_role?(course)
+    # Check if the user has an editing role in the course
+    CoursesUsers.exists?(course:, user: self, role: EDITING_ROLES)
   end
 
   EDITING_ROLES = [CoursesUsers::Roles::INSTRUCTOR_ROLE,
                    CoursesUsers::Roles::WIKI_ED_STAFF_ROLE].freeze
   def can_edit?(course)
     return true if admin?
-    return true if EDITING_ROLES.include? role(course)
+    return true if editing_role?(course)
```

Comment:
Since this concept of editing roles is only used here, I think a better option would be to take the intersection of course_roles and editing roles, like `(EDITING_ROLES & course_roles(course)).any?`, or the same `any?` pattern you are implementing in `can_see_real_names?`. No need for a separate method.